# House Prices: Advanced Regression Techniques:

## Data Cleaning & Imputation

- Eric Meyers

Source: https://www.kaggle.com/c/house-prices-advanced-regression-techniques/overview

First, import libraries.

In [1]:
import pandas as pd
import numpy as np

Add pandas display options to allow more rows/cols.

In [2]:
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 100)
pd.set_option('display.width', 1000)

Read in training data and take basic summary stats.

In [119]:
ames = pd.read_csv('./train.csv', index_col = 0)

In [120]:
ames.sample(15)

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  SalePrice
Id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            
846           85       RL          NaN    16647   Pave   NaN      IR1         Lvl    AllPub   CulDSac       Gtl       Sawyer       RRAe       Norm     1Fam     SFoyer            5            5       1975          1981     Gable  CompShg     HdBoard     HdBoard       None         0.0        TA        TA     CBlock       Gd       TA           Gd          ALQ        1390          Unf           0          0         1390    GasA        TA          Y      SBrkr      1701         0             0       1701             1             0         2         0             3             1          TA             6       Min2           2          TA    Basment       1975.0          Fin           2         611         TA         TA          Y           0            0              0          0            0         0    NaN    NaN         NaN        0       1    2007       WD        Normal     171000
1123          20       RL          NaN     8926   Pave   NaN      IR1         Lvl    AllPub    Corner       Gtl      Edwards       Norm       Norm     1Fam     1Story            4            3       1956          1956     Gable  CompShg     AsbShng     AsbShng       None         0.0        TA        TA     CBlock       TA       TA           No          Unf           0          Unf           0        672          672    GasA        Ex          Y      FuseA       960         0             0        960             0             0         1         0             3             1          TA             5        Typ           0         NaN    Basment       1956.0          Unf           1         288         TA         TA          Y          64            0              0          0          160         0    NaN  MnPrv         NaN        0      10    2009      COD       Abnorml     112000
805           20       RL         75.0     9000   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl        NAmes       Norm       Norm     1Fam     1Story            5            5       1954          1954     Gable  CompShg     Wd Sdng     Wd Sdng       None         0.0        TA        TA     CBlock       TA       TA           No          LwQ         812          Unf           0        124          936    Gas

In [121]:
ames.shape

(1460, 80)

In [122]:
ames.describe()

MSSubClass  LotFrontage        LotArea  OverallQual  OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1   BsmtFinSF2    BsmtUnfSF  TotalBsmtSF     1stFlrSF     2ndFlrSF  LowQualFinSF    GrLivArea  BsmtFullBath  BsmtHalfBath     FullBath     HalfBath  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd   Fireplaces  GarageYrBlt   GarageCars   GarageArea   WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch     PoolArea       MiscVal       MoSold       YrSold      SalePrice
count  1460.000000  1201.000000    1460.000000  1460.000000  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  1460.000000  1460.000000  1460.000000  1460.000000  1460.000000   1460.000000  1460.000000   1460.000000   1460.000000  1460.000000  1460.000000   1460.000000   1460.000000   1460.000000  1460.000000  1379.000000  1460.000000  1460.000000  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000
mean     56.897260    70.049958   10516.828082     6.099315     5.575342  1971.267808   1984.865753   103.685262   443.639726    46.549315   567.240411  1057.429452  1162.626712   346.992466      5.844521  1515.463699      0.425342      0.057534     1.565068     0.382877      2.866438      1.046575      6.517808     0.613014  1978.506164     1.767123   472.980137    94.244521    46.660274      21.954110     3.409589    15.060959     2.758904     43.489041     6.321918  2007.815753  180921.195890
std      42.300571    24.284752    9981.264932     1.382997     1.112799    30.202904     20.645407   181.066207   456.098091   161.319273   441.866955   438.705324   386.587738   436.528436     48.623081   525.480383      0.518911      0.238753     0.550916     0.502885      0.815778      0.220338      1.625393     0.644666    24.689725     0.747315   213.804841   125.338794    66.256028      61.119149    29.317331    55.757415    40.177307    496.123024     2.703626     1.328095   79442.502883
min      20.000000    21.000000    1300.000000     1.000000     1.000000  1872.000000   1950.000000     0.000000     0.000000     0.000000     0.000000     0.000000   334.000000     0.000000      0.000000   334.000000      0.000000      0.000000     0.000000     0.000000      0.000000      0.000000      2.000000     0.000000  1900.000000     0.000000     0.000000     0.000000     0.000000       0.000000     0.000000     0.000000     0.000000      0.000000     1.000000  2006.000000   34900.000000
25%      20.000000    59.000000    7553.500000     5.000000     5.000000  1954.000000   1967.000000     0.000000     0.000000     0.000000   223.000000   795.750000   882.000000     0.000000      0.000000  1129.500000      0.000000      0.000000     1.000000     0.000000      2.000000      1.000000      5.000000     0.000000  1961.000000     1.000000   334.500000     0.000000     0.000000       0.000000     0.000000     0.000000     0.000000      0.000000     5.000000  2007.000000  129975.000000
50%      50.000000    69.000000    9478.500000     6.000000     5.000000  1973.000000   1994.000000     0.000000   383.500000     0.000000   477.500000   991.500000  1087.000000     0.000000      0.000000  1464.000000      0.000000      0.000000     2.000000     0.000000      3.000000      1.000000      6.000000     1.000000  1980.000000     2.000000   480.000000     0.000000    25.000000       0.000000     0.000000     0.000000     0.000000      0.000000     6.000000  2008.000000  163000.000000
75%      70.000000    80.000000   11601.500000     7.000000     6.000000  2000.000000   2004.000000   166.000000   712.250000     0.000000   808.000000  1298.250000  1391.250000   728.000000      0.000000  1776.750000      1.000000      0.000000     2.000000     1.000000      3.000000      1.000000      7.000000     1.000000  2002.000000     2.000000   576.000000   168.000000    68.000000       0.000000     0.000000     0.000000     0.000000      0.000000     8.000000  2009.000000  214000.000000


### Data Cleaning & Imputation

First identify all NA's within all columns and get an understanding of value_counts and dtype

In [123]:
for col in ames.columns:
    print("----- %s -----" %col)
    print(ames[col].value_counts())
    print(ames[col].dtype)
    print(str(sum(pd.isna(ames[col]))) + " NA VALUES")

----- MSSubClass -----
20     536
60     299
50     144
120     87
30      69
160     63
70      60
80      58
90      52
190     30
85      20
75      16
45      12
180     10
40       4
Name: MSSubClass, dtype: int64
int64
0 NA VALUES
----- MSZoning -----
RL         1151
RM          218
FV           65
RH           16
C (all)      10
Name: MSZoning, dtype: int64
object
0 NA VALUES
----- LotFrontage -----
60.0     143
70.0      70
80.0      69
50.0      57
75.0      53
65.0      44
85.0      40
78.0      25
21.0      23
90.0      23
68.0      19
24.0      19
64.0      19
73.0      18
72.0      17
79.0      17
63.0      17
55.0      17
100.0     16
51.0      15
66.0      15
74.0      15
52.0      14
59.0      13
71.0      12
82.0      12
40.0      12
43.0      12
67.0      12
57.0      12
69.0      11
76.0      11
86.0      10
53.0      10
34.0      10
92.0      10
88.0      10
62.0       9
44.0       9
84.0       9
77.0       9
35.0       9
93.0       8
61.0       8
98.0       8
96.0 

Name: KitchenQual, dtype: int64
object
0 NA VALUES
----- TotRmsAbvGrd -----
6     402
7     329
5     275
8     187
4      97
9      75
10     47
11     18
3      17
12     11
14      1
2       1
Name: TotRmsAbvGrd, dtype: int64
int64
0 NA VALUES
----- Functional -----
Typ     1360
Min2      34
Min1      31
Mod       15
Maj1      14
Maj2       5
Sev        1
Name: Functional, dtype: int64
object
0 NA VALUES
----- Fireplaces -----
0    690
1    650
2    115
3      5
Name: Fireplaces, dtype: int64
int64
0 NA VALUES
----- FireplaceQu -----
Gd    380
TA    313
Fa     33
Ex     24
Po     20
Name: FireplaceQu, dtype: int64
object
690 NA VALUES
----- GarageType -----
Attchd     870
Detchd     387
BuiltIn     88
Basment     19
CarPort      9
2Types       6
Name: GarageType, dtype: int64
object
81 NA VALUES
----- GarageYrBlt -----
2005.0    65
2006.0    59
2004.0    53
2003.0    50
2007.0    49
1977.0    35
1998.0    31
1999.0    30
2008.0    29
1976.0    29
2000.0    27
2002.0    26
1968.0    

#### Lot Frontage

In [124]:
ames.LotFrontage.fillna(0, inplace = True) #impute 0's for 259 NA's

#### Alley

In [125]:
ames.Alley.fillna('NoAll', inplace = True) #input 'NoAll' for 1369 NA's

#### Masonry Veneer

In [126]:
ames[pd.isnull(ames.MasVnrType)] #inspect 8 NA's from 'MasVnrType' & 'MasVnrArea' to see if they match

MSSubClass MSZoning  LotFrontage  LotArea Street  Alley LotShape LandContour Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  SalePrice
Id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            
235           60       RL          0.0     7851   Pave  NoAll      Reg         Lvl    AllPub    Inside       Gtl      Gilbert       Norm       Norm     1Fam     2Story            6            5       2002          2002     Gable  CompShg     VinylSd     VinylSd        NaN         NaN        Gd        TA      PConc       Gd       TA           No          GLQ         625          Unf           0        235          860    GasA        Ex          Y      SBrkr       860      1100             0       1960             1             0         2         1             4             1          Gd             8        Typ           2          TA    BuiltIn       2002.0          Fin           2         440         TA         TA          Y         288           48              0          0            0         0    NaN   NaN         NaN        0       5    2010       WD        Normal     216500
530           20       RL          0.0    32668   Pave  NoAll      IR1         Lvl    AllPub   CulDSac       Gtl      Crawfor       Norm       Norm     1Fam     1Story            6            3       1957          1975       Hip  CompShg     Wd Sdng       Stone        NaN         NaN        Gd        TA      PConc       TA       TA           No          Rec        1219          Unf           0        816         2035    GasA        TA          Y      SBrkr      2515         0             0       2515             1             0         3         0             4             2          TA             9       Maj1           2          TA     Attchd       1975.0          RFn           2         484         TA         TA          Y           0            0            200          0            0         0    NaN   NaN         NaN        0       3    2007       WD        Alloca     200624
651           60       FV         65.0     8125   Pave  NoAll      Reg         Lvl    AllPub    Inside       Gtl      Somerst       Norm       Norm     1Fam     2Story            7            6       2007          2007     Gable  CompShg     CemntBd     CmentBd        NaN         NaN        Gd        TA      PConc       Gd       TA           No          Unf           0          Unf           0        813          813    Ga

In [127]:
ames.MasVnrType.fillna('None', inplace = True) #They do match! Impute MasVnrType to 'None'

In [128]:
ames.MasVnrArea.fillna(0, inplace = True) #Impute MasVnrArea to '0' for consistency

#### Basement

**Note** There are 37 NA's for 'BsmtQual', 'BsmtCond', 'BsmtFinType1' which are correct & consistently labeled.


**HOWEVER** 'BsmtFinType2', 'BsmtExposure' have 38 each, so there is 1 incorrect NaN each: **HOW TO HANDLE?**

In [129]:
ames[pd.isnull(ames.BsmtExposure)]

MSSubClass MSZoning  LotFrontage  LotArea Street  Alley LotShape LandContour Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  SalePrice
Id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
18            90       RL         72.0    10791   Pave  NoAll      Reg         Lvl    AllPub    Inside       Gtl       Sawyer       Norm       Norm   Duplex     1Story            4            5       1967          1967     Gable  CompShg     MetalSd     MetalSd       None         0.0        TA        TA       Slab      NaN      NaN          NaN          NaN           0          NaN           0          0            0    GasA        TA          Y      SBrkr      1296         0             0       1296             0             0         2         0             2             2          TA             6        Typ           0         NaN    CarPort       1967.0          Unf           2         516         TA         TA          Y           0            0              0          0            0         0    NaN    NaN        Shed      500      10    2006       WD        Normal      90000
40            90       RL         65.0     6040   Pave  NoAll      Reg         Lvl    AllPub    Inside       Gtl      Edwards       Norm       Norm   Duplex     1Story            4            5       1955          1955     Gable  CompShg     AsbShng     Plywood       None         0.0        TA        TA      PConc      NaN      NaN          NaN          NaN           0          NaN           0          0            0    GasA        TA          N      FuseP      1152         0             0       1152             0             0         2         0             2             2          Fa             6        Typ           0         NaN        NaN          NaN          NaN           0           0        NaN        NaN          N           0            0              0          0            0         0    NaN    NaN         NaN        0       6    2008       WD       AdjLand      82000
91            20       RL         60.0     7200   Pave  NoAll      Reg         Lvl    AllPub    Inside       Gtl        NAmes       Norm       Norm     1Fam     1Story            4            5       1950          1950     Gable  CompShg     BrkFace     Wd Sdng       None         0.0        TA        TA       Slab      NaN      NaN          NaN          NaN           0          NaN           0          0            0  

indx 949 has NaN BsmtExposure but a basement condition/qual with 0 SF. Will impute as 'No'

In [130]:
ames.loc[ames.index==949, 'BsmtExposure'] = 'No'

In [131]:
ames[pd.isnull(ames.BsmtFinType2)]

MSSubClass MSZoning  LotFrontage  LotArea Street  Alley LotShape LandContour Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  SalePrice
Id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
18            90       RL         72.0    10791   Pave  NoAll      Reg         Lvl    AllPub    Inside       Gtl       Sawyer       Norm       Norm   Duplex     1Story            4            5       1967          1967     Gable  CompShg     MetalSd     MetalSd       None         0.0        TA        TA       Slab      NaN      NaN          NaN          NaN           0          NaN           0          0            0    GasA        TA          Y      SBrkr      1296         0             0       1296             0             0         2         0             2             2          TA             6        Typ           0         NaN    CarPort       1967.0          Unf           2         516         TA         TA          Y           0            0              0          0            0         0    NaN    NaN        Shed      500      10    2006       WD        Normal      90000
40            90       RL         65.0     6040   Pave  NoAll      Reg         Lvl    AllPub    Inside       Gtl      Edwards       Norm       Norm   Duplex     1Story            4            5       1955          1955     Gable  CompShg     AsbShng     Plywood       None         0.0        TA        TA      PConc      NaN      NaN          NaN          NaN           0          NaN           0          0            0    GasA        TA          N      FuseP      1152         0             0       1152             0             0         2         0             2             2          Fa             6        Typ           0         NaN        NaN          NaN          NaN           0           0        NaN        NaN          N           0            0              0          0            0         0    NaN    NaN         NaN        0       6    2008       WD       AdjLand      82000
91            20       RL         60.0     7200   Pave  NoAll      Reg         Lvl    AllPub    Inside       Gtl        NAmes       Norm       Norm     1Fam     1Story            4            5       1950          1950     Gable  CompShg     BrkFace     Wd Sdng       None         0.0        TA        TA       Slab      NaN      NaN          NaN          NaN           0          NaN           0          0            0  

In [132]:
ames.loc[ames.index == 333]

MSSubClass MSZoning  LotFrontage  LotArea Street  Alley LotShape LandContour Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  SalePrice
Id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
333          20       RL         85.0    10655   Pave  NoAll      IR1         Lvl    AllPub    Inside       Gtl      NridgHt       Norm       Norm     1Fam     1Story            8            5       2003          2004     Gable  CompShg     VinylSd     VinylSd    BrkFace       296.0        Gd        TA      PConc       Gd       TA           No          GLQ        1124          NaN         479       1603         3206    GasA        Ex          Y      SBrkr      1629         0             0       1629             1             0         2         0             3             1          Gd             7        Typ           1          Gd     Attchd       2003.0          RFn           3         880         TA         TA          Y           0            0              0          0            0         0    NaN   NaN         NaN        0      10    2009       WD        Normal     284000

indx 333 has NaN BsmtFinType2, but other potentially dependent features:
    - Foundation: 'PConc'
    - BsmtExposure: 'No'
    - BsmtFinType1: 'GLQ'
Let's isolate and group other like-rows to hopefully find a pattern in BsmtFinType2

In [133]:
Bsmt_grp = ames.groupby(['Foundation', 'BsmtExposure','BsmtFinType1', 'BsmtFinType2'])

In [134]:
Bsmt_grp.count()

MSSubClass  MSZoning  LotFrontage  LotArea  Street  Alley  LotShape  LandContour  Utilities  LotConfig  LandSlope  Neighborhood  Condition1  Condition2  BldgType  HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd  RoofStyle  RoofMatl  Exterior1st  Exterior2nd  MasVnrType  MasVnrArea  ExterQual  ExterCond  BsmtQual  BsmtCond  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  Heating  HeatingQC  CentralAir  Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  KitchenQual  TotRmsAbvGrd  Functional  Fireplaces  FireplaceQu  GarageType  GarageYrBlt  GarageFinish  GarageCars  GarageArea  GarageQual  GarageCond  PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  PoolQC  Fence  MiscFeature  MiscVal  MoSold  YrSold  SaleType  SaleCondition  SalePrice
Foundation BsmtExposure BsmtFinType1 BsmtFinType2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              
BrkTil     Av           ALQ          Unf                    1         1            1        1       1      1         1            1          1          1          1             1           1           1         1           1            1            1          1             1          1         1            1            1           1           1          1          1         1         1           1           1          1            1        1          1           1           1         1         1             1          1             1             1         1         1             1             1            1             1           1           1            0           1            1             1           1           1           1           1           1           1            1              1          1            1         1       0      0            0        1       1       1         1              1          1
           Gd           BLQ          Unf                    1         1            1        1       1      1         1            1          1          1          1             1           1           1         1           1            1            1          1             1          1         1            1            1           1           1          1          1         1         1           1           1          1            1        1          1           1           1         1         1             1          1             1             1         1         1             1             1            1             1           1           1            0           0            0             0           1           1           0           0           1           1            1              1          1            1         1       0      0            0        1       1       1         1              1          1
           Mn           ALQ          Unf                    1         1            1        1       1      1         1            1          1          1          1             1           1           1         1           1            1            1          1             1          1         1            1            1           1     

Of basements with Foundation: 'PConc', BsmtExposure: 'No', and BsmtFinType1: 'GLQ'; 149 of 153 BsmtFinType2's are 'Unf'. PLUS BsmtUnfSF = 1603. Will impute idx 333 as 'Unf'.

In [135]:
ames.loc[ames.index == 333, 'BsmtFinType2'] = 'Rec'

Now let's impute the other Bsmt values that genuinely have no basement as 'NB' (No Basement)

In [136]:
BsmtNaNs = ames[ames['BsmtFinType2'].isnull()].index

In [137]:
ames.loc[BsmtNaNs, 'BsmtQual'] = 'NB'
ames.loc[BsmtNaNs, 'BsmtCond'] = 'NB'
ames.loc[BsmtNaNs, 'BsmtExposure'] = 'NB'
ames.loc[BsmtNaNs, 'BsmtFinType1'] = 'NB'
ames.loc[BsmtNaNs, 'BsmtFinType2'] = 'NB'

#### Electrical

**Note** There is one NA in Electrical. **HOW TO HANDLE?**

In [138]:
ames[pd.isnull(ames.Electrical)]

MSSubClass MSZoning  LotFrontage  LotArea Street  Alley LotShape LandContour Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  SalePrice
Id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            
1380          80       RL         73.0     9735   Pave  NoAll      Reg         Lvl    AllPub    Inside       Gtl       Timber       Norm       Norm     1Fam       SLvl            5            5       2006          2007     Gable  CompShg     VinylSd     VinylSd       None         0.0        TA        TA      PConc       Gd       TA           No          Unf           0          Unf           0        384          384    GasA        Gd          Y        NaN       754       640             0       1394             0             0         2         1             3             1          Gd             7        Typ           0         NaN    BuiltIn       2007.0          Fin           2         400         TA         TA          Y         100            0              0          0            0         0    NaN   NaN         NaN        0       5    2008       WD        Normal     167500

In [139]:
ames['Electrical'].value_counts()

SBrkr    1334
FuseA      94
FuseF      27
FuseP       3
Mix         1
Name: Electrical, dtype: int64

Let's see the counts of combinations on potentially Electrical-related variables.

In [140]:
Elec_group = ames.groupby(['Heating', 'CentralAir', 'Electrical'])

In [141]:
Elec_group.count()

MSSubClass  MSZoning  LotFrontage  LotArea  Street  Alley  LotShape  LandContour  Utilities  LotConfig  LandSlope  Neighborhood  Condition1  Condition2  BldgType  HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd  RoofStyle  RoofMatl  Exterior1st  Exterior2nd  MasVnrType  MasVnrArea  ExterQual  ExterCond  Foundation  BsmtQual  BsmtCond  BsmtExposure  BsmtFinType1  BsmtFinSF1  BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  HeatingQC  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  KitchenQual  TotRmsAbvGrd  Functional  Fireplaces  FireplaceQu  GarageType  GarageYrBlt  GarageFinish  GarageCars  GarageArea  GarageQual  GarageCond  PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  PoolQC  Fence  MiscFeature  MiscVal  MoSold  YrSold  SaleType  SaleCondition  SalePrice
Heating CentralAir Electrical                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
Floor   N          SBrkr                1         1            1        1       1      1         1            1          1          1          1             1           1           1         1           1            1            1          1             1          1         1            1            1           1           1          1          1           1         1         1             1             1           1             1           1          1            1          1         1         1             1          1             1             1         1         1             1             1            1             1           1           1            0           1            1             1           1           1           1           1           1           1            1              1          1            1         1       0      0            0        1       1       1         1              1          1
GasA    N          FuseA               16        16           16       16      16     16        16           16         16         16         16            16          16          16        16          16           16           16         16            16         16        16           16           16          16          16         16         16          16        16        16            16            16          16            16          16         16           16         16        16        16            16         16            16            16        16        16            16            16           16            16          16          16            1          10           10            10          16          16          10          10          16          16           16             16         16           16        16       0      3            1       16      16      16        16             16         16
                   FuseF               12        12           12       12      12     12        12           12         12         12         12            12          12          12        12          12           12           12         12            12         12        12           12           12          12 

Most common Electrical type with Heating as 'GasA' & CentralAir as 'Y' is 'SBrkr' by far. We will impute this.

In [142]:
ames.loc[ames.index == 1380,'Electrical'] = 'SBrkr'

#### Checkpoint: What's left?

In [143]:
ames.isnull().sum()

MSSubClass          0
MSZoning            0
LotFrontage         0
LotArea             0
Street              0
Alley               0
LotShape            0
LandContour         0
Utilities           0
LotConfig           0
LandSlope           0
Neighborhood        0
Condition1          0
Condition2          0
BldgType            0
HouseStyle          0
OverallQual         0
OverallCond         0
YearBuilt           0
YearRemodAdd        0
RoofStyle           0
RoofMatl            0
Exterior1st         0
Exterior2nd         0
MasVnrType          0
MasVnrArea          0
ExterQual           0
ExterCond           0
Foundation          0
BsmtQual            0
BsmtCond            0
BsmtExposure        0
BsmtFinType1        0
BsmtFinSF1          0
BsmtFinType2        0
BsmtFinSF2          0
BsmtUnfSF           0
TotalBsmtSF         0
Heating             0
HeatingQC           0
CentralAir          0
Electrical          0
1stFlrSF            0
2ndFlrSF            0
LowQualFinSF        0
GrLivArea 

**Checkpoint Note**: Fireplace (690) NA's, Garage (81) NA's, Pool (1453) NA's, Fence (1179) NA's, & MiscFeature (1406) NA's.

#### Fireplace

In [144]:
ames[['FireplaceQu', 'Fireplaces']][ames.FireplaceQu.isnull()]

FireplaceQu  Fireplaces
Id                          
1            NaN           0
6            NaN           0
11           NaN           0
13           NaN           0
16           NaN           0
18           NaN           0
19           NaN           0
20           NaN           0
27           NaN           0
30           NaN           0
31           NaN           0
32           NaN           0
33           NaN           0
37           NaN           0
39           NaN           0
40           NaN           0
43           NaN           0
44           NaN           0
45           NaN           0
48           NaN           0
49           NaN           0
50           NaN           0
51           NaN           0
53           NaN           0
57           NaN           0
58           NaN           0
60           NaN           0
61           NaN           0
62           NaN           0
64           NaN           0
65           NaN           0
68           NaN           0
69           NaN           0
72           NaN           0
74           NaN           0
75           NaN           0
76           NaN           0
77           NaN           0
78           NaN           0
79           NaN           0
80           NaN           0
82           NaN           0
84           NaN           0
88           NaN           0
89           NaN           0
90           NaN           0
91           NaN           0
92           NaN           0
93           NaN           0
95           NaN           0
97           NaN           0
98           NaN           0
99           NaN           0
100          NaN           0
103          NaN           0
104          NaN           0
107          NaN           0
108          NaN           0
109          NaN           0
111          NaN           0
118          NaN           0
122          NaN           0
123          NaN           0
124          NaN           0
126          NaN           0
130          NaN           0
133          NaN           0
134          NaN           0
138          NaN           0
140          NaN           0
142          NaN           0
143          NaN           0
144          NaN           0
145          NaN           0
146          NaN           0
147          NaN           0
149          NaN           0
150          NaN           0
151          NaN           0
155          NaN           0
156          NaN           0
157          NaN           0
161          NaN           0
164          NaN           0
165          NaN           0
166          NaN           0
171          NaN           0
180          NaN           0
184          NaN           0
185          NaN           0
187          NaN           0
188          NaN           0
192          NaN           0
193          NaN           0
194          NaN           0
195          NaN           0
199          NaN           0
201          NaN           0
203          NaN           0
205          NaN           0
206          NaN           0
210          NaN           0
211          NaN           0
212          NaN           0
214          NaN           0
215          NaN           0
217          NaN           0
218          NaN           0
220          NaN           0
221          NaN           0
224          NaN           0
226          NaN           0
228          NaN           0
231          NaN           0
234          NaN           0
236          NaN           0
237          NaN           0
238          NaN           0
239          NaN           0
241          NaN           0
242          NaN           0
243          NaN           0
247          NaN           0
249          NaN           0
251          NaN           0
253          NaN           0
254          NaN           0
255          NaN           0
257          NaN           0
260          NaN           0
264          NaN           0
265          NaN           0
275          NaN           0
276          NaN           0
277          NaN           0
278          NaN           0
282

These are genuinely null. Impute as 'NF' (No Fireplace)

In [145]:
ames.loc[ames['FireplaceQu'].isnull(), 'FireplaceQu'] = 'NF'

#### Garage

In [146]:
ames[ames['GarageType'].isnull()]

MSSubClass MSZoning  LotFrontage  LotArea Street  Alley LotShape LandContour Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  SalePrice
Id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
40            90       RL         65.0     6040   Pave  NoAll      Reg         Lvl    AllPub    Inside       Gtl      Edwards       Norm       Norm   Duplex     1Story            4            5       1955          1955     Gable  CompShg     AsbShng     Plywood       None         0.0        TA        TA      PConc       NB       NB           NB           NB           0           NB           0          0            0    GasA        TA          N      FuseP      1152         0             0       1152             0             0         2         0             2             2          Fa             6        Typ           0          NF        NaN          NaN          NaN           0           0        NaN        NaN          N           0            0              0          0            0         0    NaN    NaN         NaN        0       6    2008       WD       AdjLand      82000
49           190       RM         33.0     4456   Pave  NoAll      Reg         Lvl    AllPub    Inside       Gtl      OldTown       Norm       Norm   2fmCon     2Story            4            5       1920          2008     Gable  CompShg     MetalSd     MetalSd       None         0.0        TA        TA     BrkTil       TA       TA           No          Unf           0          Unf           0        736          736    GasA        Gd          Y      SBrkr       736       716             0       1452             0             0         2         0             2             3          TA             8        Typ           0          NF        NaN          NaN          NaN           0           0        NaN        NaN          N           0            0            102          0            0         0    NaN    NaN         NaN        0       6    2009      New       Partial     113000
79            90       RL         72.0    10778   Pave  NoAll      Reg         Lvl    AllPub    Inside       Gtl       Sawyer       Norm       Norm   Duplex     1Story            4            5       1968          1968       Hip  CompShg     HdBoard     HdBoard       None         0.0        TA        TA     CBlock       TA       TA           No          Unf           0          Unf           0       1768         1768  

GarageType,GarageYrBlt,GarageFinish,GarageQual,GarageCond are all (81) consistently NaN with 0 Garage Area. Impute 'NG' for No Garage.

In [147]:
GrgNans = ames[ames['GarageType'].isnull()].index

In [148]:
ames.loc[GrgNans, 'GarageType'] = 'NG'
ames.loc[GrgNans, 'GarageYrBlt'] = 'NG'
ames.loc[GrgNans, 'GarageFinish'] = 'NG'
ames.loc[GrgNans, 'GarageQual'] = 'NG'
ames.loc[GrgNans, 'GarageCond'] = 'NG'

**Future Note** may have trouble with 'GarageYrBlt' later in formatting dates.

#### Pool

In [149]:
ames[['PoolQC', 'PoolArea']][ames['PoolQC'].isnull()]

PoolQC  PoolArea
Id                   
1       NaN         0
2       NaN         0
3       NaN         0
4       NaN         0
5       NaN         0
6       NaN         0
7       NaN         0
8       NaN         0
9       NaN         0
10      NaN         0
11      NaN         0
12      NaN         0
13      NaN         0
14      NaN         0
15      NaN         0
16      NaN         0
17      NaN         0
18      NaN         0
19      NaN         0
20      NaN         0
21      NaN         0
22      NaN         0
23      NaN         0
24      NaN         0
25      NaN         0
26      NaN         0
27      NaN         0
28      NaN         0
29      NaN         0
30      NaN         0
31      NaN         0
32      NaN         0
33      NaN         0
34      NaN         0
35      NaN         0
36      NaN         0
37      NaN         0
38      NaN         0
39      NaN         0
40      NaN         0
41      NaN         0
42      NaN         0
43      NaN         0
44      NaN         0
45      NaN         0
46      NaN         0
47      NaN         0
48      NaN         0
49      NaN         0
50      NaN         0
51      NaN         0
52      NaN         0
53      NaN         0
54      NaN         0
55      NaN         0
56      NaN         0
57      NaN         0
58      NaN         0
59      NaN         0
60      NaN         0
61      NaN         0
62      NaN         0
63      NaN         0
64      NaN         0
65      NaN         0
66      NaN         0
67      NaN         0
68      NaN         0
69      NaN         0
70      NaN         0
71      NaN         0
72      NaN         0
73      NaN         0
74      NaN         0
75      NaN         0
76      NaN         0
77      NaN         0
78      NaN         0
79      NaN         0
80      NaN         0
81      NaN         0
82      NaN         0
83      NaN         0
84      NaN         0
85      NaN         0
86      NaN         0
87      NaN         0
88      NaN         0
89      NaN         0
90      NaN         0
91      NaN         0
92      NaN         0
93      NaN         0
94      NaN         0
95      NaN         0
96      NaN         0
97      NaN         0
98      NaN         0
99      NaN         0
100     NaN         0
101     NaN         0
102     NaN         0
103     NaN         0
104     NaN         0
105     NaN         0
106     NaN         0
107     NaN         0
108     NaN         0
109     NaN         0
110     NaN         0
111     NaN         0
112     NaN         0
113     NaN         0
114     NaN         0
115     NaN         0
116     NaN         0
117     NaN         0
118     NaN         0
119     NaN         0
120     NaN         0
121     NaN         0
122     NaN         0
123     NaN         0
124     NaN         0
125     NaN         0
126     NaN         0
127     NaN         0
128     NaN         0
129     NaN         0
130     NaN         0
131     NaN         0
132     NaN         0
133     NaN         0
134     NaN         0
135     NaN         0
136     NaN         0
137     NaN         0
138     NaN         0
139     NaN         0
140     NaN         0
141     NaN         0
142     NaN         0
143     NaN         0
144     NaN         0
145     NaN         0
146     NaN         0
147     NaN         0
148     NaN         0
149     NaN         0
150     NaN         0
151     NaN         0
152     NaN         0
153     NaN         0
154     NaN         0
155     NaN         0
156     NaN         0
157     NaN         0
158     NaN         0
159     NaN         0
160     NaN         0
161     NaN         0
162     NaN         0
163     NaN         0
164     NaN         0
165     NaN         0
166     NaN         0
167     NaN         0
168     NaN         0
169     NaN         0
170     NaN         0
171     NaN         0
172     NaN         0
173     NaN         0
174     NaN         0
175     NaN         0
176     NaN         0
177     NaN         0
178     NaN         0
179     NaN         0
180     NaN         0
1

PoolQC NaN's are truely No Pool. Impute with 'NP'.

In [150]:
ames.loc[ames.PoolQC.isnull(), 'PoolQC'] = 'NP'

#### Fence

In [151]:
ames[ames.Fence.isnull()]

MSSubClass MSZoning  LotFrontage  LotArea Street  Alley LotShape LandContour Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu GarageType GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  SalePrice
Id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
1             60       RL         65.0     8450   Pave  NoAll      Reg         Lvl    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam     2Story            7            5       2003          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace       196.0        Gd        TA      PConc       Gd       TA           No          GLQ         706          Unf           0        150          856    GasA        Ex          Y      SBrkr       856       854             0       1710             1             0         2         1             3             1          Gd             8        Typ           0          NF     Attchd        2003          RFn           2         548         TA         TA          Y           0           61              0          0            0         0     NP   NaN         NaN        0       2    2008       WD        Normal     208500
2             20       RL         80.0     9600   Pave  NoAll      Reg         Lvl    AllPub       FR2       Gtl      Veenker      Feedr       Norm     1Fam     1Story            6            8       1976          1976     Gable  CompShg     MetalSd     MetalSd       None         0.0        TA        TA     CBlock       Gd       TA           Gd          ALQ         978          Unf           0        284         1262    GasA        Ex          Y      SBrkr      1262         0             0       1262             0             1         2         0             3             1          TA             6        Typ           1          TA     Attchd        1976          RFn           2         460         TA         TA          Y         298            0              0          0            0         0     NP   NaN         NaN        0       5    2007       WD        Normal     181500
3             60       RL         68.0    11250   Pave  NoAll      IR1         Lvl    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam     2Story            7            5       2001          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace       162.0        Gd        TA      PConc       Gd       TA           Mn          GLQ         486          Unf           0        434          920    GasA  

From description file: NA in 'Fence' means No Fence. No potentially dependent variables to check against. Impute to 'NF' for No Fence.

In [154]:
ames.loc[ames.Fence.isnull(), 'Fence'] = 'NF'

#### MiscFeature

If MiscFeature's are truely NA, they will have MiscVal of $0. Let's check.

In [162]:
ames[['MiscFeature', 'MiscVal']][ames.MiscFeature.isnull()]

MiscFeature  MiscVal
Id                       
1            NaN        0
2            NaN        0
3            NaN        0
4            NaN        0
5            NaN        0
7            NaN        0
9            NaN        0
10           NaN        0
11           NaN        0
12           NaN        0
13           NaN        0
14           NaN        0
15           NaN        0
16           NaN        0
19           NaN        0
20           NaN        0
21           NaN        0
22           NaN        0
23           NaN        0
24           NaN        0
25           NaN        0
26           NaN        0
27           NaN        0
28           NaN        0
29           NaN        0
30           NaN        0
31           NaN        0
32           NaN        0
33           NaN        0
34           NaN        0
35           NaN        0
36           NaN        0
37           NaN        0
38           NaN        0
39           NaN        0
40           NaN        0
41           NaN        0
42           NaN        0
43           NaN        0
44           NaN        0
45           NaN        0
46           NaN        0
47           NaN        0
48           NaN        0
49           NaN        0
50           NaN        0
51           NaN        0
53           NaN        0
54           NaN        0
55           NaN        0
56           NaN        0
57           NaN        0
58           NaN        0
59           NaN        0
60           NaN        0
61           NaN        0
62           NaN        0
63           NaN        0
64           NaN        0
65           NaN        0
66           NaN        0
67           NaN        0
68           NaN        0
69           NaN        0
70           NaN        0
71           NaN        0
72           NaN        0
73           NaN        0
74           NaN        0
75           NaN        0
76           NaN        0
77           NaN        0
78           NaN        0
79           NaN        0
80           NaN        0
81           NaN        0
82           NaN        0
83           NaN        0
84           NaN        0
86           NaN        0
87           NaN        0
88           NaN        0
89           NaN        0
90           NaN        0
91           NaN        0
92           NaN        0
93           NaN        0
94           NaN        0
95           NaN        0
97           NaN        0
98           NaN        0
101          NaN        0
102          NaN        0
103          NaN        0
104          NaN        0
105          NaN        0
106          NaN        0
108          NaN        0
109          NaN        0
110          NaN        0
111          NaN        0
112          NaN        0
113          NaN        0
114          NaN        0
115          NaN        0
116          NaN        0
117          NaN        0
118          NaN        0
119          NaN        0
120          NaN        0
121          NaN        0
122          NaN        0
123          NaN        0
124          NaN        0
125          NaN        0
126          NaN        0
127          NaN        0
128          NaN        0
129          NaN        0
130          NaN        0
131          NaN        0
132          NaN        0
133          NaN        0
134          NaN        0
135          NaN        0
136          NaN        0
137          NaN        0
138          NaN        0
139          NaN        0
140          NaN        0
141          NaN        0
142          NaN        0
143          NaN        0
144          NaN        0
145          NaN        0
146          NaN        0
147          NaN        0
148          NaN        0
149          NaN        0
150          NaN        0
151          NaN        0
152          NaN        0
153          NaN        0
154          NaN        0
155          NaN        0
156          NaN        0
157          NaN        0
158          NaN        0
159          NaN        0
160          NaN        0
161          NaN        0
162          NaN        0
1

Confirmed MiscVal of 0 for all NA MiscFeatures. Impute as NMF (No Misc Features).

In [164]:
ames.loc[ames.MiscFeature.isnull(), 'MiscFeature'] = 'NMF'

In [166]:
ames.isnull().sum()

MSSubClass       0
MSZoning         0
LotFrontage      0
LotArea          0
Street           0
Alley            0
LotShape         0
LandContour      0
Utilities        0
LotConfig        0
LandSlope        0
Neighborhood     0
Condition1       0
Condition2       0
BldgType         0
HouseStyle       0
OverallQual      0
OverallCond      0
YearBuilt        0
YearRemodAdd     0
RoofStyle        0
RoofMatl         0
Exterior1st      0
Exterior2nd      0
MasVnrType       0
MasVnrArea       0
ExterQual        0
ExterCond        0
Foundation       0
BsmtQual         0
BsmtCond         0
BsmtExposure     0
BsmtFinType1     0
BsmtFinSF1       0
BsmtFinType2     0
BsmtFinSF2       0
BsmtUnfSF        0
TotalBsmtSF      0
Heating          0
HeatingQC        0
CentralAir       0
Electrical       0
1stFlrSF         0
2ndFlrSF         0
LowQualFinSF     0
GrLivArea        0
BsmtFullBath     0
BsmtHalfBath     0
FullBath         0
HalfBath         0
BedroomAbvGr     0
KitchenAbvGr     0
KitchenQual 

## Imputation Complete!